corehq/apps/ota/utils.py

```diff
@@ -0,0 +1,68 @@
+from django.utils.translation import ugettext as _
+from casexml.apps.case.xml import V2
+from casexml.apps.phone.restore import RestoreConfig, RestoreParams
+from corehq.apps.domain.models import Domain
+from dimagi.utils.logging import notify_exception
+
+from .models import DemoUserRestore
+
+
+def turn_off_demo_mode(commcare_user):
+    """
+    Turns demo mode OFF for commcare_user and deletes existing demo restore
+    """
+
+    # delete old restore
+    old_restore_id = commcare_user.demo_restore_id
+    if old_restore_id:
+        old_restore = DemoUserRestore.objects.get(id=old_restore_id)
+        old_restore.delete()
+
+    commcare_user.demo_restore_id = None
+    commcare_user.is_demo_user = False
+    commcare_user.save()
+
+
+def turn_on_demo_mode(commcare_user, domain):
+    """
+    Turns demo mode ON for commcare_user, and resets restore to latest
+    """
+    try:
+        # the order of following two is important, because the restore XML should contain...
+        # user_type='demo' in user_data element
+        commcare_user.is_demo_user = True
+        reset_demo_user_restore(commcare_user, domain)
+    except Exception as e:
+        notify_exception(None, message=e.message)
+        return {'errors': [
+            _("Something went wrong in creating restore for the user. Please try again or report an issue")
+        ]}
+    else:
+        commcare_user.save()
+        return {'errors': []}
+
+
+def reset_demo_user_restore(commcare_user, domain):
+    """
+    Updates demo restore for the demo commcare_user
+    """
+    assert commcare_user.domain == domain
+
+    # get latest restore
+    restore = RestoreConfig(
+        project=Domain.get_by_name(domain),
+        user=commcare_user.to_casexml_user(),
+        params=RestoreParams(version=V2),
+    ).get_payload().as_file()
+    demo_restore = DemoUserRestore.create(commcare_user._id, restore)
```

Comment:
I don't see anywhere in the call path for this function where an existing DemoUserRestore get's deleted so I think this will result in duplicate objects. Is that intentional?